### ISSUE TITLE
Large file support on i386 arch seems missing?

### ISSUE THREAD

**Issue by ktmf01** [google/oss-fuzz#10311]
Today I came across a rather strange bug: https://oss-fuzz.com/testcase-detail/6275951100297216

During debugging, I found out that it seems large file support is silently absent somehow. The code assumes it is present, causing the fuzzer to get stuck in an infinite loop. Problem in the code is here: https://github.com/xiph/flac/blob/c65ef58924938e2375c427d0cfca6931d278f95c/src/flac/encode.c#L2943

In the testcase, this function fskip_ahead is called with 4294964736 as the offset argument. off_t has a width of 8 bytes, which I verified by casting to off_t, the value stays the same. However, for some reason, the code seeks backward. At the start of the function, ftell(f) returns 13384, at the end it returns 10824. This means it seeked back 2560 bytes, which is what I get if I cast 4294964736 to int32_t.

It seems to me fseeko is somehow accepting a 64-bit int value, but internally wraps it to 32-bit int? Or am I missing something? Could this be caused by the instrumentation?

**Comment by Alan32Liu** [comment]
Requesting help from @jonathanmetzman.

**Comment by jonathanmetzman** [comment]
>Could this be caused by the instrumentation?

Maybe try building the fuzzer with --sanitizer none and seeing if it repros (you can definitely build centipede targets with no sanitizer).

**Comment by jonathanmetzman** [comment]
Other than that, I'd say i386 is sort of a best effort offering and not a big priority, if you have a fix we can make, we'd be happy to accept it. Sorry

**Comment by ktmf01** [comment]
> Other than that, I'd say i386 is sort of a best effort offering and not a big priority, if you have a fix we can make, we'd be happy to accept it. Sorry

Yes, I understand that. I posted this here hoped it would be something obvious, but it isn't. I was just trying to get rid of all the fuzz finds, but after looking into this, I think it is safe to ignore. Feel free to ignore the ramblings below, I'd like to keep it around for later reference

-----

I've tried a bunch of builds
- i386 ASAN reproduces (timeout, > 60 seconds) 
- i386 none works fine (84ms)
- i386 UBSAN works fine (80ms)
- i386 MSAN fails to build
- i386 TSAN fails to build
- x86_64 ASAN works fine (428ms). However, https://oss-fuzz.com/testcase-detail/6275951100297216 says x86_64 ASAN does in fact reproduce, it does not when I try locally
- x86_64 UBSAN works fine (399ms). However, https://oss-fuzz.com/testcase-detail/6275951100297216 says x86_64 UBSAN is 'flaky'.

I've tried debugging. If I build with no fuzzer (on i386) I ftello(f) returns 13384 on line 2941 and 8589934591 on line 2944. If I build with ASAN, I get 13384 on line 2941 and 10824 on line 2944. They're both calling fseeko64 in /lib32/libc.so.6, so there is no interceptor.

It seems this is somehow a compiler bug. When build without a sanitizer, the registers are as follows upon entering fseeko64
```
(gdb) info registers
eax            0xffffa690          -22896
ecx            0x8eac980           149604736
edx            0xfffff600          -2560
ebx            0xffffcd1c          -13028
esp            0xffffa68c          0xffffa68c
ebp            0xffffa758          0xffffa758
esi            0x0                 0
edi            0x1                 1
eip            0xf7cc0270          0xf7cc0270 <fseeko64>
eflags         0x246               [ PF ZF IF ]
cs             0x23                35
ss             0x2b                43
ds             0x2b                43
es             0x2b                43
fs             0x0                 0
gs             0x63                99
```

This makes sense, according [to this](https://stackoverflow.com/questions/2535989/what-are-the-calling-conventions-for-unix-linux-system-calls-and-user-space-f) parameters for syscalls on i386 go into registers, and %edx and %esi can be concatenated to form the off_t operand. I see no other possible combination. The assembly of fskip_ahead agrees:

```
   0x08134a90 <+224>:   call   0x808c450 <__sanitizer_cov_trace_const_cmp4()>
   0x08134a95 <+229>:   mov    -0x98(%ebp),%eax
   0x08134a9b <+235>:   cmp    $0x8000,%eax
   0x08134aa0 <+240>:   je     0x8134ab7 <fskip_ahead+263>
   0x08134aa6 <+246>:   mov    0x82e59b2,%al
   0x08134aab <+251>:   add    $0x1,%al
   0x08134aad <+253>:   mov    %al,0x82e59b2
   0x08134ab2 <+258>:   jmp    0x8134b27 <fskip_ahead+375>
   0x08134ab7 <+263>:   mov    -0xc(%ebp),%ecx
   0x08134aba <+266>:   mov    -0x18(%ebp),%edx
   0x08134abd <+269>:   mov    -0x14(%ebp),%esi
   0x08134ac0 <+272>:   mov    %esp,%eax
   0x08134ac2 <+274>:   mov    %esi,0x8(%eax)
   0x08134ac5 <+277>:   mov    %edx,0x4(%eax)
   0x08134ac8 <+280>:   mov    %ecx,(%eax)
   0x08134aca <+282>:   movl   $0x1,0xc(%eax)
   0x08134ad1 <+289>:   call   0x80554d0 <fseeko64@plt>
```
So, whatever is pointed to by the stack pointer gets split up in edx and esi. Seems to make sense

However, when building with ASAN, the registers at the start of fseeko64 contain something else
```
(gdb) info registers
eax            0xffff82d0          -32048
ecx            0xf2e00d60          -220197536
edx            0xfffff600          -2560
ebx            0xf2e00d60          -220197536
esp            0xffff82cc          0xffff82cc
ebp            0xffff8448          0xffff8448
esi            0xffff8380          -31872
edi            0x0                 0
eip            0xf7cc0270          0xf7cc0270 <fseeko64>
eflags         0x282               [ SF IF ]
cs             0x23                35
ss             0x2b                43
ds             0x2b                43
es             0x2b                43
fs             0x0                 0
gs             0x63                99
```
esi doesn't contain zero, but it doesn't contain 0xffffffff either, so I'm not entirely sure I'm understanding things right. However, the assembly calling fseeko64 is

```
   0x08214bb1 <+497>:   call   0x80a1f20 <__sanitizer_cov_trace_const_cmp4()>
   0x08214bb6 <+502>:   mov    0x48(%esi),%eax
   0x08214bb9 <+505>:   add    $0x10,%esp
   0x08214bbc <+508>:   cmp    $0x8000,%eax
   0x08214bc1 <+513>:   je     0x8214bd8 <fskip_ahead+536>
   0x08214bc7 <+519>:   mov    0x855e21c,%al
   0x08214bcc <+524>:   add    $0x1,%al
   0x08214bce <+526>:   mov    %al,0x855e21c
   0x08214bd3 <+531>:   jmp    0x8214cc7 <fskip_ahead+775>
   0x08214bd8 <+536>:   mov    0xa8(%esi),%ecx
   0x08214bde <+542>:   mov    0xa0(%esi),%edx
   0x08214be4 <+548>:   mov    0xa4(%esi),%edi
   0x08214bea <+554>:   sub    $0x10,%esp
   0x08214bed <+557>:   mov    %esp,%eax
   0x08214bef <+559>:   mov    %edi,0x8(%eax)
   0x08214bf2 <+562>:   mov    %edx,0x4(%eax)
   0x08214bf5 <+565>:   mov    %ecx,(%eax)
   0x08214bf7 <+567>:   movl   $0x1,0xc(%eax)
   0x08214bfe <+574>:   call   0x806b2f0 <fseeko64@plt>
```

Somehow this code is different, not splitting the operand over edx and eci but edx and edi instead.

**Comment by jonathanmetzman** [comment]
Yep, ASAN on i386 is aggressively unsupported because Google doesn't use i386 for anything (except chrome)
